[Exporter/OpenSearch] OpenSearch exporter setup

MitchellGale opened the pull request:
**Description:**
Adding initial set-up for OpenSearch exporter addition. Future PRs will include adding functionality to the exporter, code coverage and e2e tests. 

Broken up for easier review.

[New component proposal.](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/23611)

**Link to tracking Issue:**
Start of resolution of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7905.

**Testing:**
Will come in future PR to 80+% coverage.


Break-up of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23045

MitchellGale committed:
Added space

MitchellGale committed:
Added MitchellGale to CODEOWNERS

MitchellGale committed:
Added OpenSearchExporter builder-config.yaml.

MitchellGale committed:
Added OpenSearchExporter go.mod otel contrib col.

MitchellGale committed:
Added MakeFile for OpenSearchExporter.

MitchellGale committed:
Added OpenSearchExporter for versions.yaml

MitchellGale committed:
Added OpenSearchExporter for .golangci.yml, bug_report.yaml, feature_request.yaml, other.yaml

MitchellGale committed:
Reverted changes in go.mod

MitchellGale committed:
Reverted MakeFile space.

MitchellGale committed:
Reverted go.mod manual change.

MitchellGale committed:
added OS config.go

MitchellGale committed:
added link for OS config.go

MitchellGale committed:
Added go mod

MitchellGale committed:
Renamed Makefile

MitchellGale committed:
Switched go mod go version from 1.20 to 1.19

MitchellGale committed:
Ran `make genotelcontribcol`

MitchellGale committed:
Made with go mod changes and adding factory.go, and adding config to config.go

MitchellGale committed:
Fixed issues in config and factor go.

MitchellGale committed:
go sum

MitchellGale committed:
Added YangDB and Max Ksyunz to code owners for OpenSearch exporter.

MitchellGale committed:
Merge pull request #20 from Bit-Quill/mainCopy

OpenSearch Exporter initial PR

codeboten commented:
Pinging @Aneurysm9 for the initial review

MitchellGale committed:
Removed commented out code. Added @Aneurysm9 as code owner.

Aneurysm9 commented on cmd/otelcontribcol/builder-config.yaml:
```diff
@@ -66,6 +66,7 @@ exporters:
   - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/lokiexporter v0.80.0
   - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/mezmoexporter v0.80.0
   - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/opencensusexporter v0.80.0
+  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/opensearchexporter v0.80.0
```
I don't believe this should be added here at this time.  Per the [new component contribution guidelines](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md):

>* **Last PR** should mark the new component as `Alpha` stability and add it to the `cmd/otelcontribcol`
  binary by updating the `cmd/otelcontribcol/components.go` file. The component must be enabled
  only after sufficient testing and only when it meets [`Alpha` stability requirements](https://github.com/open-telemetry/opentelemetry-collector#alpha).

MitchellGale replied:
Thank you, it has been removed.

Aneurysm9 commented on exporter/opensearchexporter/config.go:
```diff
@@ -0,0 +1,7 @@
+
+package opensearchexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/opensearchexporter"
+
+type Config struct {
```
This should have configuration structure at this point.  It may be wrong and need to change in the future, but we should have an idea at this point what the configuration should look like.  There should also be a README that outlines the component's purpose and usage, referring to the configuration as appropriate.

MitchellGale replied:
Added content for configuration structure.

Aneurysm9 commented on exporter/opensearchexporter/factory.go:
```diff
@@ -0,0 +1,13 @@
+
+// NewFactory creates a factory for OpenSearch exporter.
+func NewFactory() exporter.Factory {
+	return nil
```
This needs to return an actual component factory.  Even if the handler functions drop data on the floor, it should be possible after this PR to include the component in a build and wire it into a pipeline.

MitchellGale replied:
Made change to add component factory.

MitchellGale committed:
Removed opensearch from builder-config.yaml

Adding config.go content for OpenSearch exporter.

Adding new factory.go

Adding new sso model

OS bulk, readme and adding content to new factory.

Removed opensearch_bulk.go, sso_model.go, sso_trace_exporter.go, trace_exporter.go. sorted imports in factor.go.

MitchellGale committed:
Merge remote-tracking branch 'upstream/main' into Integ/OpenSearchExporter-setup

MitchellGale committed:
Updated README.md

MitchellGale committed:
Updated go mod and go sum

MitchellGale committed:
Added change log for opensearch.

MitchellGale committed:
Updated dependabot.yml

MitchellGale committed:
Reverted changes in otelcontrib.

MitchellGale committed:
Reverted changes in versions.

MitchellGale committed:
Added new line at end of new component

MitchellGale committed:
Reverted file to golangci.yml

MitchellGale committed:
Added Opensearch exporter to excluded modules in versions.yaml.

Aneurysm9 commented on .github/dependabot.yml:
```diff
@@ -1097,8 +1102,3 @@ updates:
       interval: "weekly"
       day: "wednesday"
-  - package-ecosystem: "gomod"
-    directory: "/receiver/syslogreceiver"
```
This doesn't seem correct.  I don't think any modules should be removed here.

MitchellGale replied:
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19410 I think it's to do with this issue. Is there a work-around that you recommend?

MaxKsyunz replied:
Without this change, [check gendependebot](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/5489576612/jobs/10003924866?pr=23819) action in the check workflow will fail. 

I thought it'd be best to err on the side of passing CI. There are about 5 components already out of dependabot.yml due to the same problem.

Aneurysm9 commented on exporter/opensearchexporter/README.md:
```diff
@@ -0,0 +1,66 @@
+
+## Configuration options
+
+- `endpoints`: List of OpenSearch URLs. If endpoints is missing, the
+  OPENSEARCH_URL environment variable will be used.
```
This is not typically how configuration defaults are handled in the collector.  As environment variables can be interpolated into configuration files it is usually left to the user to determine what variables to use.  This will be complicated by the fact that this takes a list.  How would the value of this variable be parsed to obtain a list of URLs?  What will the exporter do with a list of endpoints?  Does it export simultaneously to all endpoints? Load-balance between them in some way?

MaxKsyunz replied:
Fair point re: environment variable.

As for `endpoints` property, it is passed to the OpenSearch Go client. It maintains it's own connection pool.

Aneurysm9 commented on exporter/opensearchexporter/README.md:
```diff
@@ -0,0 +1,66 @@
+
+- `endpoints`: List of OpenSearch URLs. If endpoints is missing, the
+  OPENSEARCH_URL environment variable will be used.
+- `num_workers` (optional): Number of workers publishing bulk requests concurrently.
```
Is there a default to note here?  Presumably `1` but it would be good to call out here.  It's also odd that this is noted as `optional` while other values have their default noted, which presumably also means they're optional.

MaxKsyunz replied:
This value is passed to the OpenSearch Go client as well. It defaults to the number of CPUs.

Aneurysm9 replied:
Would be good to link to the details about how the default is determined, then.  I presume it's actually `GOMAXPROCS` under the hood, which may be an important distinction for some users.

Aneurysm9 commented on exporter/opensearchexporter/README.md:
```diff
@@ -0,0 +1,66 @@
+  OPENSEARCH_URL environment variable will be used.
+- `num_workers` (optional): Number of workers publishing bulk requests concurrently.
+- `dataset` (default=`default`) a user-provided label. It is used to construct the name of the destination index or data stream.
+- `namespace` (default=`namespace`)a user-provided label. It is used to construct the name of the destination index or data stream.
```
Is `namespace` a sensible default value here?  It feels odd but I'm not super familiar with typical OpenSearch usage.  If namespaces are optional then a default of an empty string might make more sense, but I'll have to defer to your expertise here.

MaxKsyunz replied:
This is following the specification [here](https://github.com/opensearch-project/opensearch-catalog/blob/main/docs/schema/observability/Naming-convention.md).

@YANG-DB  Does default namespace value of `namespace` crucial?

YANG-DB replied:
Hi
yes - without any other specific index name we fallback to this naming convention 
we have an index pattern for all signal types to catch these index names:
 - `ss4o_logs-*-*`
 - `ss4o_traces-*-*`
 - `ss4o_metrics-*-*`

Aneurysm9 commented on exporter/opensearchexporter/README.md:
```diff
@@ -0,0 +1,66 @@
+- `num_workers` (optional): Number of workers publishing bulk requests concurrently.
+- `dataset` (default=`default`) a user-provided label. It is used to construct the name of the destination index or data stream.
+- `namespace` (default=`namespace`)a user-provided label. It is used to construct the name of the destination index or data stream.
+- `flush`: Event bulk buffer flush settings
+    - `bytes` (default=5242880): Write buffer flush limit.
+    - `interval` (default=30s): Write buffer time limit.
```
We [generally recommend](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#general-recommendations) against introducing any batching within an exporter.  This is better handled via existing processors that can handle batching telemetry data.

MaxKsyunz replied:
These are settings internal to the OpenSearch client. 

The exporter may create several JSON documents for one `pdata.Traces` record.  These settings control how these documents are sent to OpenSearch. The exporter does not do any batching of records it receives from the collector.

Aneurysm9 replied:
Why the `interval` value, then?  I can understand having a maximum export data size requiring splitting an incoming record into multiple outgoing requests, but the `interval` value implies to me that the exporter will hold onto incoming records for some period of time before sending them if the maximum data size is not yet reached.

MaxKsyunz replied:
Yes, current exporter implementation maintains one bulk indexer object from opensearch-go client which, in turn maintains this queue.

I believe it's reasonable to create bulk indexer for each `pushTraceData` call and make this setting unnecessary.

Aneurysm9 commented on exporter/opensearchexporter/README.md:
```diff
@@ -0,0 +1,66 @@
+- `flush`: Event bulk buffer flush settings
+    - `bytes` (default=5242880): Write buffer flush limit.
+    - `interval` (default=30s): Write buffer time limit.
+- `retry`: Event retry settings
```
This should probably use the [`exporterhelper.RetrySettings` type](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/queued_retry.go#L239) to provide consistency of configuration across exporters.

MaxKsyunz replied:
These values are used to create a backoff function for the OpenSearch client. 
We'll check if `exporterhelper.RetrySettings` can be used here.

MaxKsyunz replied:
The config struct can be changed to use `exporterhelper.RetrySettings` but the implied behaviour may be different.

`exporterhelper.WithRetry` retries a batch of trace records, right? OpenSearch supports retrying one failed document from a bulk indexing request.

Would it be more appropriate to log documents that OpenSearch failed to index and move on? Currently they are added back to the bulk indexer queue to be tried later.

Aneurysm9 replied:
You can use the [`consumererror.Traces` type](https://github.com/open-telemetry/opentelemetry-collector/blob/main/consumer/consumererror/signalerrors.go#L27-L29) to return an error containing the spans that failed to export.  The `exportererror` [error handler](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/traces.go#L50-L56) will unwrap the error, extract the spans, and replace the request with one containing just the failed spans.

Aneurysm9 commented on exporter/opensearchexporter/README.md:
```diff
@@ -0,0 +1,66 @@
+    - `initial_interval` (default=100ms): Initial waiting time if a HTTP request failed.
+    - `max_interval` (default=1m): Max waiting time if a HTTP request failed.
+
+### HTTP settings
```
This should use the [`confighttp.HTTPClientSettings` struct](https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/confighttp/confighttp.go#L31) to provide consistency of configuration and functionality across exporters.  Using `confighttp` can cover for the HTTP, [TLS](https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configtls/configtls.go#L65), and AuthN settings included here.

MaxKsyunz replied:
Yes, I think this is feasible.

Aneurysm9 commented on exporter/opensearchexporter/README.md:
```diff
@@ -0,0 +1,66 @@
+- `timeout` (default=90s): HTTP request time limit.
+- `headers` (optional): Headers to be send with each HTTP request.
+
+### Security and Authentication settings
```
Rather than directly implementing HTTP basic auth it would be better to use the [`configauth` package](https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configauth/README.md) (which is [included in](https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/confighttp/confighttp.go#L56) the `confighttp.HTTPClientSettings` mentioned above) which can be used to provide a [`basicauthextension`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/basicauthextension) reference or any other client authentication extension.  This will provide significantly more flexibility to users.

MaxKsyunz replied:
These settings are passed to OpenSearch Go client. I'll check if it's possible to use `configauth` here instead.

Aneurysm9 replied:
Ideally the exporter would be able to provide an `http.Client` to the OpenSearch Go client.  If that's not possible at this time would it be possible to expose such affordances in the OpenSearch Go client library?

MaxKsyunz replied:
At the moment, opensearch-go manages HTTP client internally.

Exporter could use settings from `configauth` and initialize opensearch client from those. Would that make sense?Refactoring opensearch-go client is a larger project.

Aneurysm9 commented on exporter/opensearchexporter/factory.go:
```diff
@@ -0,0 +1,58 @@
+		HTTPClientSettings: HTTPClientSettings{
+			Timeout: 90 * time.Second,
+		},
+		Namespace: "",
+		Dataset:   "",
```
These don't appear to align with the defaults documented in the README.

MitchellGale replied:
Good catch. Updated and fixed in e240c46aca23e87a1420d36e615d2e5aec6e84bc.

MitchellGale committed:
Adding defaults for Namespace and Dataset in factory.go as outlined in OpenSearch readme.

MitchellGale committed:
swapped defaults for Namespace and Dataset in factory.go as outlined in OpenSearch readme.

MitchellGale committed:
Merge remote-tracking branch 'upstream/main' into Integ/OpenSearchExporter-setup

MitchellGale committed:
Update dependabot.

MitchellGale committed:
Update README.md with details on default num_workers

MaxKsyunz commented:
@Aneurysm9 based on your feedback and how opensearch-go client is set up, I'd like to do the following:
1. Use `HTTPClientSettings`  in exporter's configuration. Map these settings to opensearch-go client's config.
2. Limit OpenSearch client to one BulkIndexer worker. A worker will be created and shut down for each batch of trace data.
3. Use `RetrySettings` but retry only if the whole batch fails.

This way the exporter will not use internal queue management done by opensearch-go and use existing opentel-collector data structures and code where possible.

What do you think?

MaxKsyunz committed:
Address review feedback.

Remove exporter-specific settings and use confighttp.HTTPClientSettings and exporterhelper.RetrySettings

Update README to reflect the above.

Signed-off-by: Max Ksyunz <max.ksyunz@improving.com>

Aneurysm9 reviewed (approved):
LGTM.

I've still got some questions around the retry and authN logic, but I think we can deal with those on the implementation PRs when they're more concrete conversations.

MaxKsyunz committed:
Re-run fmt, mod tidy

Signed-off-by: Max Ksyunz <max.ksyunz@improving.com>

MaxKsyunz committed:
Merge branch 'main' into integ/OpenSearchExporter-setup

# Conflicts:
#	.github/dependabot.yml

MaxKsyunz committed:
Re-run gendependabot

dependabot.yml reached maximum limit so a package is getting dropped but CI will fail without this change.

See #19410

Signed-off-by: Max Ksyunz <max.ksyunz@improving.com>

MaxKsyunz commented:
GHAs failed on a known flaky test. See https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21094

Is there a way to re-run it?

fullammo commented:
Hi! can we enable this exporter for logs as well?

steevex35 commented:
Hi! 
This exporter can be used for logs ?
the OpensearchExporter readme.md mentions only trace pipeline. 
thank you
